### ISSUE TITLE
Box::into_raw(Box::from(&[1,2,3])) causes unexpected memory deallocation that compiler does not complain

### ISSUE THREAD

**Issue by hanabi1224** [rust-lang/rust#53054]
I have a test case that when I add '&' by mistake, compiler does not complain but it causes early memory deallocation thus test fails

**Correct**:

```rust
let features1_ptr = Box::into_raw(Box::from([
            Box::into_raw(Box::from([100.0, 40.0, 22.0])) as *const f64,
            Box::into_raw(Box::from([211.0, 20.0, 2.0]))as *const f64,
            Box::into_raw(Box::from([32.0, 190.0, 150.0]))as *const f64,
            Box::into_raw(Box::from([2.0, 100.0, 100.0]))as *const f64,
        ])) as *const *const f64;
```

**Problematic**:

```rust
let features1_ptr = Box::into_raw(Box::from([
            Box::into_raw(Box::from(&[100.0, 40.0, 22.0])) as *const f64,
            Box::into_raw(Box::from(&[211.0, 20.0, 2.0]))as *const f64,
            Box::into_raw(Box::from(&[32.0, 190.0, 150.0]))as *const f64,
            Box::into_raw(Box::from(&[2.0, 100.0, 100.0]))as *const f64,
        ])) as *const *const f64;
```

Code to repro:
https://github.com/hanabi1224/RuFastEMD/blob/401d86343af23813b28989336fe557f8657dca00/tests/test.rs#L55

**Comment by xfix** [comment]
Why should it complain? Yes, you probably shouldn't cast `*mut &[f64; 3]` into `*const f64`, but as far the compiler is concerned, you can use `as` to cast a pointer type to any other pointer type. This is an advanced feature, an user is trusted to use it safely.

Something you may find useful is that slices have `as_ptr` method which allows you to get a pointer to a slice without casting.

```rust
pub fn x(callback: fn(*const *const f64)) {
    let a = [100.0, 40.0, 22.0];
    let b = [211.0, 20.0, 2.0];
    let c = [32.0, 190.0, 150.0];
    let d = [2.0, 100.0, 100.0];
    let features = [a.as_ptr(), b.as_ptr(), c.as_ptr(), d.as_ptr()];
    callback(features.as_ptr());
}
```

Note the separate assignments. Rust won't help you with lifetimes when you use raw pointers.

**Comment by hanabi1224** [comment]
@xfix  This is a test case for ffi api, and all I want is a unmanaged pointer so that I can manually free it later, , I used .as_ptr() at first but it causes early memory deallocation as well and broke the test case.

**Comment by sfackler** [comment]
Nothing is being deallocated early. In the broken example you're making a `Box<&[f64; 3]>`, not a `Box<[f64; 3]>`. There's an extra pointer.

**Comment by xfix** [comment]
By the way, you may want to use `Box::leak(Box::new([100.0, 40.0, 22.0])).as_ptr()` instead.

**Comment by hanabi1224** [comment]
@xfix  thx for the trick, that's much better!

**Comment by hanabi1224** [comment]
@sfackler  Then I don't quite understand why below are both inferred as *const i32

![image](https://user-images.githubusercontent.com/6546635/43732996-0e318ab8-9968-11e8-8f6b-8b20a049e12e.png)

**Comment by sfackler** [comment]
`as_ptr()` isn't a method on `Box`, it's a method on `&[f32]`.

**Comment by hanabi1224** [comment]
@sfackler  that makes sense , thx!
